crates/uv/src/commands/pip/tree.rs

```diff
@@ -496,6 +501,125 @@ impl<'env> DisplayDependencyGraph<'env> {
         lines
     }
 
+    /// Aggregate the requirements associated with the incoming edges for a node.
+    fn aggregate_requirement(&self, cursor: &Cursor) -> Option<String> {
+        let (url, specifiers) = cursor.edge_ids().iter().fold(
+            (None::<String>, Vec::new()),
+            |(url, mut specs), edge_id| {
+                let requirement = &self.graph[*edge_id];
+
+                let url = match requirement.version_or_url.as_ref() {
+                    None => url,
+                    Some(VersionOrUrl::VersionSpecifier(values)) => {
+                        specs.extend(values.iter().cloned());
+                        url
+                    }
+                    Some(VersionOrUrl::Url(value)) => url.or_else(|| Some(value.to_string())),
+                };
+
+                (url, specs)
+            },
+        );
+
+        if let Some(url) = url {
+            return Some(url);
+        }
+
+        if specifiers.is_empty() {
+            return None;
+        }
+
+        let display = Self::simplify_specifiers(specifiers).to_string();
+
+        if display.is_empty() {
+            None
+        } else {
+            Some(display)
+        }
+    }
+
+    /// Simplify a collection of specifiers into a canonical representation for display.
+    fn simplify_specifiers(specifiers: Vec<VersionSpecifier>) -> VersionSpecifiers {
+        let (lower, upper, others) = specifiers.into_iter().fold(
+            (None, None, Vec::new()),
+            |(lower, upper, mut rest), spec| match *spec.operator() {
+                Operator::GreaterThan | Operator::GreaterThanEqual => {
+                    (Some(Self::prefer_lower(lower, spec)), upper, rest)
+                }
+                Operator::LessThan | Operator::LessThanEqual => {
+                    (lower, Some(Self::prefer_upper(upper, spec)), rest)
+                }
+                _ => {
+                    rest.push(spec);
+                    (lower, upper, rest)
+                }
+            },
+        );
+
+        let mut merged = lower
+            .into_iter()
+            .chain(upper)
+            .chain(others)
+            .collect::<Vec<_>>();
+
+        let mut seen = FxHashSet::default();
+
+        merged.retain(|spec| seen.insert(spec.to_string()));
+
+        VersionSpecifiers::from_iter(merged)
```

Comment:
I'm a little wary of this. We almost certainly already have logic for combining version specifiers, though I'm not sure we can use it here, e.g.

https://github.com/astral-sh/uv/blob/3f83390e342dcb6fccc449335c15d1255885ad20/crates/uv-pep440/src/version_ranges.rs#L13-L22

It does look pretty clean though and is nicer than emitting redundant ranges, e.g., `>=3.6, >=3.7`.

@konstin might have some things.

Reply:
From what I understand, this is a lossy conversion, e.g. once we intersect the specifiers into intervals for PubGrub we lose the exact shape of the original constraints, and there isn’t a reliable way to turn an arbitrary range back into a finite set of PEP 440 specifiers.

We could instead move the specifier-level deduplication that tightens bounds into the PEP 440 crate, exposing for other potential use-cases and decluttering the `pip tree` implementation. What do you think?